consensus/validator/validator.go

```diff
@@ -0,0 +1,198 @@
+	}
+
+	// Get set of txns that we can execute
+	start, end := a.latestExecutedTxn, a.latestExecutedTxn
+	for a.txnPool[end+1].Transaction != nil {
+		end++
+	}
+
+	txnsToExecute := make([]mempool.BroadcastedTransaction, end-start+1)
+	for i := range txnsToExecute {
+		txnsToExecute[i] = mempool.BroadcastedTransaction{
+			Transaction:   txns[start].Transaction,
+			DeclaredClass: txns[start].Class,
+			PaidFeeOnL1:   txns[start].PaidFeeOnL1,
+		}
+	}
+
+	if err := a.builder.ExecuteTxns(txnsToExecute); err != nil {
+		return err
+	}
+	a.latestExecutedTxn = end + 1
+	return nil
+}
+
+// ProposalCommitment checks the set of proposed commitments against those generated locally.
+func (a *validator[V, H, A]) ProposalCommitment(proCom *types.ProposalCommitment) error {
+	commitments, concatCount, err := a.builder.ExecutePending()
+	if err != nil {
+		return err
+	}
+	pendingBlock := a.builder.PendingBlock()
+
+	if err := compareProposalCommitment(proCom, pendingBlock.Header, commitments, concatCount); err != nil {
+		return err
+	}
+	return nil
+}
+
+// ProposalFin executes the provided transactions, and stores the result in the pending state
+func (a *validator[V, H, A]) ProposalFin(proposalFin types.ProposalFin) error {
+	pendingBlock := a.builder.PendingBlock()
+
+	proposerCommitmentFelt := felt.Felt(proposalFin)
+	if !proposerCommitmentFelt.Equal(pendingBlock.Hash) {
+		return fmt.Errorf("proposal fin: commitments do not match: expected=%s actual=%s",
+			proposerCommitmentFelt.String(), pendingBlock.Hash.String())
+	}
+	return a.builder.Finalise(nil)
+}
+
+// Todo: there are fields in ProposalCommitment that we don't check against. Some of these fields
+// will be dropped in the finalised spec, so I don't think we should worry about them until then
+//  1. Some fields we can't get / compute: VersionConstantCommitment, NextL2GasPriceFRI
+//  2. The gas prices. Currently the spec sets eth gas prices, but in v1, these will be dropped
+//     for fri prices.
+func compareProposalCommitment( //nolint:gocyclo // disable linter because this function has a simple pattern
+	p *types.ProposalCommitment,
+	h *core.Header,
+	c *core.BlockCommitments,
+	concatCount *felt.Felt,
+) error {
+	compareProposalField := func(name string, a, b *felt.Felt) error {
+		if a.IsZero() && b == nil {
+			return nil
+		}
+		if b != nil && a.Equal(b) {
+			return nil
+		}
+		return fmt.Errorf("%s commitment mismatch: proposal=%s commitments=%s", name, a, b)
+	}
+
+	if p.BlockNumber != h.Number {
+		return fmt.Errorf("block number mismatch: proposal=%d header=%d", p.BlockNumber, h.Number)
+	}
+
+	if !p.ParentCommitment.Equal(h.ParentHash) {
+		return fmt.Errorf("parent hash mismatch: proposal=%s header=%s", p.ParentCommitment.String(), h.ParentHash.String())
+	}
+
+	if !p.Builder.Equal(h.SequencerAddress) {
+		return fmt.Errorf("builder mismatch: proposal=%s header=%s", p.Builder.String(), h.SequencerAddress.String())
+	}
+
+	if p.Timestamp > h.Timestamp {
```

Comment:
Why do we have this comparison? I'm wondering if there is any rule for the timestamp.